File: R/tableaux.R
.asciiTableau <- function(tableau){
  .C("asciiTableauR", tableau=tableau, l=length(tableau), result="")$result
}

.dualTableau <- function(tableau){
  .C("dualTableauR", tableau=tableau, l=length(tableau), result=list(0L))$result[[1L]]
}

.isPartition <- function(partition){
  .C("isPartitionR", partition=partition, l=length(partition), result=0L)$result
}

.countStandardYoungTableaux <- function(shape){
  .C("countStandardYoungTableauxR", partition=shape, l=length(shape),
     result=0L)$result
}

.standardYoungTableaux <- function(shape){
  .C("standardYoungTableauxR", partition=shape, l=length(shape),
     result=list(0L))$result[[1L]]
}

.hookLengths <- function(shape){
  .C("hookLengthsR", partition=shape, l=length(shape), result=list(0L))$result[[1L]]
}

#' ASCII diagram of a tableau
#'
#' Returns the ASCII diagram of a tableau.
#'
#' @param tableau a tableau
#'
#' @return A character string.
#' @export
#' @useDynLib tableaux
#'
#' @examples
#' tableau <- list(c(1,3,4,6,7), c(2,5,10,8), 9)
#' cat(asciiTableau(tableau))
asciiTableau <- function(tableau){
  .asciiTableau(lapply(tableau, as.integer))
}

#' Dual tableau
#'
#' Returns the dual tableau (mirror image to the main diagonal).
#'
#' @param tableau a tableau
#'
#' @return A tableau.
#' @export
#'
#' @examples
#' tableau <- list(c(1,3,4,6,7), c(2,5,10,8), 9)
#' cat(asciiTableau(tableau))
#' cat(asciiTableau(dualTableau(tableau)))
dualTableau <- function(tableau){
  .dualTableau(lapply(tableau, as.integer))
}

#' Number of standard Young tableaux
#'
#' Count the number of standard Young tableaux of a given shape.
#'
#' @param shape partition of integers
#'
#' @return An integer.
#' @export
#' @importFrom purrr map_int
#'
#' @examples
#' countStandardYoungTableaux(c(5,3,1))
countStandardYoungTableaux <- function(shape){
  shape <- purrr::map_int(shape, as.integer)
  if(.isPartition(shape)==0L){
    warning("`shape` is not a partition")
  }
  .countStandardYoungTableaux(shape)
}

#' Standard Young tableaux
#'
#' Returns the list of standard Young tableaux of a given shape.
#'
#' @param shape partition of integers
#'
#' @return A list of tableaux
#' @export
#'
#' @examples
#' shape <- c(3,2)
#' standardYoungTableaux(shape)
standardYoungTableaux <- function(shape){
  shape <- purrr::map_int(shape, as.integer)
  if(.isPartition(shape)==0L){
    warning("`shape` is not a partition")
  }
  .standardYoungTableaux(shape)
}

#' Hook lengths
#'
#' Hook lengths of a partition.
#'
#' @param shape a partition
#'
#' @return The tableau of hook lengths.
#' @export
#' @importFrom purrr map_int
#'
#' @examples
#' hookLengths(c(5,2,1))
hookLengths <- function(shape){
  shape <- purrr::map_int(shape, as.integer)
  if(.isPartition(shape)==0L){
    warning("`shape` is not a partition")
  }
  .hookLengths(shape)
}

File: inst/Haskell/src/Tableaux.hs
{-# LANGUAGE DataKinds                #-}
{-# LANGUAGE ForeignFunctionInterface #-}
module Tableaux where
import           Data.Bool                        (bool)
import           Data.Singletons                  (sing)
import qualified Data.Vector.SEXP                 as VS
import           Foreign
import           Foreign.C
import           Foreign.R                        (SEXP)
import qualified Foreign.R.Type                   as R
import           Language.R.Literal               (mkProtectedSEXPVectorIO)
import           Math.Combinat.Partitions.Integer
import           Math.Combinat.Tableaux

importTableau :: Ptr (SEXP s R.Int) -> Ptr CInt -> IO (Tableau Int32)
importTableau tableau l = do
  l <- peek l
  tableau <- peekArray (fromIntegral l :: Int) tableau
  return $ map (VS.toList . VS.fromSEXP) tableau

foreign export ccall asciiTableauR :: Ptr (SEXP s R.Int) -> Ptr CInt -> Ptr CString -> IO ()
asciiTableauR :: Ptr (SEXP s R.Int) -> Ptr CInt -> Ptr CString -> IO ()
asciiTableauR tableau l result = do
  tableau <- importTableau tableau l
  let ascii = show (asciiTableau tableau)
  (>>=) (newCString ascii) (poke result)

foreign export ccall dualTableauR :: Ptr (SEXP s R.Int) -> Ptr CInt ->
                                                  Ptr (SEXP s R.Vector) -> IO ()
dualTableauR :: Ptr (SEXP s R.Int) -> Ptr CInt -> Ptr (SEXP s R.Vector) -> IO ()
dualTableauR tableau l result = do
  tableau <- importTableau tableau l
  dtableau <- mkProtectedSEXPVectorIO sing $
    (map (VS.toSEXP . VS.fromList) (dualTableau tableau) :: [SEXP s R.Int])
  poke result dtableau

foreign export ccall isPartitionR :: Ptr CInt -> Ptr CInt -> Ptr CInt -> IO()
isPartitionR :: Ptr CInt -> Ptr CInt -> Ptr CInt -> IO()
isPartitionR partition l result = do
  l <- peek l
  partition <- peekArray (fromIntegral l :: Int) partition
  poke result $ bool 0 1 (isPartition (map fromIntegral partition :: [Int]))

foreign export ccall countStandardYoungTableauxR :: Ptr CInt -> Ptr CInt ->
                                                                Ptr CInt -> IO()
countStandardYoungTableauxR :: Ptr CInt -> Ptr CInt -> Ptr CInt -> IO()
countStandardYoungTableauxR partition l result = do
  l <- peek l
  partition <- peekArray (fromIntegral l :: Int) partition
  poke result $ fromIntegral $
                  countStandardYoungTableaux (mkPartition $
                                                map fromIntegral partition)

foreign export ccall standardYoungTableauxR :: Ptr CInt -> Ptr CInt ->
                                                   Ptr (SEXP s R.Vector) -> IO()
standardYoungTableauxR :: Ptr CInt -> Ptr CInt -> Ptr (SEXP s R.Vector) -> IO()
standardYoungTableauxR partition l result = do
  l <- peek l
  partition <- peekArray (fromIntegral l :: Int) partition
  let tableaux = standardYoungTableaux (mkPartition $ map fromIntegral partition)
  let tableaux32 = map (map (map fromIntegral)) tableaux :: [[[Int32]]]
  tableauxAsSEXP <- mapM (\x -> (mkProtectedSEXPVectorIO sing $
                            (map (VS.toSEXP . VS.fromList) x :: [SEXP s R.Int]))
                              :: IO (SEXP s R.Vector)) tableaux32
  (>>=) (mkProtectedSEXPVectorIO sing tableauxAsSEXP) (poke result)

foreign export ccall hookLengthsR :: Ptr CInt -> Ptr CInt ->
                                                  Ptr (SEXP s R.Vector) -> IO ()
hookLengthsR :: Ptr CInt -> Ptr CInt -> Ptr (SEXP s R.Vector) -> IO ()
hookLengthsR partition l result = do
  l <- peek l
  partition <- peekArray (fromIntegral l :: Int) partition
  let lengths = map (map fromIntegral) $
                  hookLengths (mkPartition $ map fromIntegral partition)
                    :: [[Int32]]
  (>>=) (mkProtectedSEXPVectorIO sing $
          (map (VS.toSEXP . VS.fromList) lengths :: [SEXP s R.Int]))
            (poke result)
